### ISSUE TITLE
Variables defined in base classes in incompatible way despite `Any`

### ISSUE THREAD

**Issue by vberlier** [microsoft/pyright#3225]
I'm extending a generic base class and providing `Any` as a type parameter because instances of this class are used in a context where I only care about the extra functionality. The problem is that Pyright reports that members using the type parameter are incompatible even though `Any` explicitly says that any type is fine.

```python
T = TypeVar("T")

@dataclass
class Base(Generic[T]):
    foo: Union[T, None] = None

class BaseWithExtra(Base[Any]):
    def extra_method(self):
        print(self.foo)

# Cannot create consistent method ordering
# Base classes for class "Derived" define variable "foo" in incompatible way
# Base class "BaseWithExtra" provides type "str | None"
# Base class "Base[str]" provides type "Any | None"
class Derived(Base[str], BaseWithExtra):
    pass

l: List[BaseWithExtra] = []
l.append(Derived("hello"))

for obj in l:
    obj.extra_method()
```

**Comment by erictraut** [comment]
If you run this code in the Python interpreter, you'll receive a runtime exception because there is no consistent MRO (method resolution order) ordering. That's because you're deriving a class from the same base class twice, which isn't allowed.

Here's the workaround I recommend. This works both at runtime and type checks without error.

```python
@dataclass
class Base(Generic[T]):
    foo: Union[T, None] = None


class BaseWithExtra(Base[T]):
    def extra_method(self):
        print(self.foo)


class Derived(BaseWithExtra[str]):
    pass
```

**Comment by vberlier** [comment]
Couldn't get back to this until now but it seems I over-simplified when reproducing the error. My use-case is actually closer to the following:

```python
@dataclass
class Base(Generic[T]):
    foo: Union[T, None] = None

class TextBase(Base[str]):
    pass

class BinaryBase(Base[bytes]):
    pass

class BaseWithExtra(Base[Any]):
    def extra_method(self):
        print(self.foo)

# Cannot create consistent method ordering
# Base classes for class "Derived" define variable "foo" in incompatible way
# Base class "BaseWithExtra" provides type "str | None"
# Base class "TextBase" provides type "Any | None"
class Derived(TextBase, BaseWithExtra):
    pass

l: List[BaseWithExtra] = []
l.append(Derived("hello"))

for obj in l:
    obj.extra_method()
```

This works in the interpreter and outputs `hello`. The `Base` is supposed to be abstract and the library exposes specialized subclasses for different types (`TextBase`, `BinaryBase`). `BaseWithExtra` is a marker and implements extra functionality for making the class work for a specific use-case. `BaseWithExtra` really only cares about the `Base` class, it works the same for all specialized subclasses.

In this case multiple inheritance works around what would be an `n * m` problem where each concrete `XBase` would need its own `XBaseWithExtra`. So the issue isn't with MRO but the fact that `Any` is treated as an incompatible type in this context.

**Comment by erictraut** [comment]
Yes, this is intended behavior. You are enabling a strict-mode check `reportIncompatibleVariableOverride` which validates that classes that are combined in a multiple-inheritance hierarchy have matching types.

If you want to leave this diagnostic check enabled, here is the proper way to annotate your classes.

```python
class BaseWithExtra(Base[T]):
    def extra_method(self):
        print(self.foo)


class Derived(TextBase, BaseWithExtra[str]):
    pass
```

If you don't like the way this diagnostic check works, you have the option of leaving it disabled.
